src/util/emitUtils.js

```diff
@@ -578,22 +580,29 @@ function emitToClients(io, nsp, rooms, key, obj) {
  */
 function doEmit(nsp, key, obj) {
   const newObjectAsString = getNewObjAsString(key, obj); // { key: {new: obj }}
-  const numClients = Object.values(nsp.connected).length;
-  if (key.startsWith('refocus.internal.realtime.sample')) {
-    tracker.trackEmit(obj.name, obj.updatedAt, numClients);
-  }
   pubSubStats.trackEmit(key, obj);
-  if (!toggle.isFeatureEnabled('enableClientStats')) { // dont enter the else if its not a sample
-    nsp.emit(key, newObjectAsString);
-  } else if (key.startsWith('refocus.internal.realtime.sample')) {
+  if (toggle.isFeatureEnabled('enableClientStats') && key.startsWith('refocus.internal.realtime.sample')) { // dont enter the else if its not a sample
+    let numClients = 0;
+    const rooms = nsp.rooms.slice(0);
+    nsp.rooms = [];
     Object.values(nsp.connected)
+      .filter((socket) => rooms.includes(Object.keys(socket.rooms)[1]))
       .forEach((socket) => {
+        numClients++;
         socket.emit(key, newObjectAsString, (time) => {
           // wrap this around to check the key and include for only samples
           tracker.trackClient(obj.name, obj.updatedAt, time);
           pubSubStats.trackClient(key, obj, time)
         });
       });
+      tracker.trackEmit(obj.name, obj.updatedAt, numClients);
+  } else {
+    const numClients = nsp.rooms.length ? Object.values(nsp.connected)
+    .filter((socket) => nsp.rooms.includes(Object.keys(socket.rooms)[1])).length : Object.keys(nsp.connected).length;
```

Comment:
Oh yeah, you're right, we do need this for both. In that case, probably better to do it in one place above like you were doing before - I only suggested moving it into the with-callbacks block because I was thinking it was only needed there.

You could combine this with my above comment and set a sockets variable above, then use that for both the iteration and the numClients.

Reply:
Though I'm not sure I like having to filter all the sockets for every emit - could have a performance impact.
Maybe the call to trackEmit should only include numClients if the toggle is enabled?
Yeah, I think the only reason to track the number of clients emitted to is to have something to compare to the number of clients received.

So actually I think you should just remove this from the else and call trackEmit without the numClients arg. (and update trackEmit to allow this).

@iamigo thoughts?